created multiple views for the equities page

kayfoh opened the pull request.

kayfoh committed:
created multiple views for the equities page

kayfoh committed:
replace constant tag reference with parameters. created event_listener and series_data_processor functions

dswired commented:
Still reviewing but can you always put your references in the description? Thanks

dswired commented on fin-advisor-crm/markets/views.py:
```diff
@@ -1,47 +1,239 @@
+EQUITY_DB_PRICE_COL = "closing_price"
+
+
+def format_currency(value):
+    if value >= 1_000_000:
+        return "GHS{:,.0f} M".format(value / 1_000_000)
+    elif value >= 1_000:
+        return "GHS{:,.0f} K".format(value / 1_000)
+    else:
+        return "GHS{:,.0f}".format(value)
```
Best to put this in `core/utils.py` so its available for use throughout the application.

dswired commented on fin-advisor-crm/markets/views.py:
```diff
@@ -1,47 +1,239 @@
+    return JsonResponse({'unique_tickers': unique_tickers})
+
+
+def get_ticker_info(share_code):
+    ticker_url = f'https://dev.kwayisi.org/apis/gse/equities/{share_code}'
+    ticker_pull = requests.get(ticker_url)
+    if ticker_pull.ok:
+        return ticker_pull.json()
+    else:
+        raise ValueError
+
+
+def get_ticker_info_basic(share_code):
+    ticker_url = f"https://dev.kwayisi.org/apis/gse/live/{share_code}"
+    ticker_pull = requests.get(ticker_url)
+    if ticker_pull.ok:
+        return ticker_pull.json()
+    else:
+        raise ValueError
```
Both of these functions have the same signature so you need just one function with the `url` and `share code` as inputs.

dswired commented on fin-advisor-crm/markets/views.py:
```diff
@@ -1,47 +1,239 @@
+    return per_ticker_data     
+
+
+def equity_snap_data_pull(
+    selected_ticker, 
+    as_of_date
+    ):
+    if as_of_date:
+        as_of_date = datetime.datetime.strptime(
+            as_of_date, 
+            "%Y-%m-%d"
+        ).date()
+        last_record = DailyPublicEquityTrades.objects.filter(
+            ticker=selected_ticker, 
+            trade_date=as_of_date
+        ).latest("trade_date")
+    else:
+        last_record = DailyPublicEquityTrades.objects.filter(
+            ticker=selected_ticker
+        ).latest("trade_date")
+    return last_record
```
Here it seems you should be calling `equity_trend_data_pull` and then calling `.latest("trade_date")` on the result. No need to rewrite a function that does this.

dswired commented on fin-advisor-crm/markets/views.py:
```diff
@@ -1,47 +1,239 @@
+        raise ValueError
+
+
+def equity_trend_data_pull(
```
I would give this a more descriptive name like `get_equity_trend_data`.

dswired commented on fin-advisor-crm/markets/views.py:
```diff
@@ -1,47 +1,239 @@
+    return last_record
+
+
+def event_listener(request):
```
Why this name?

dswired commented on fin-advisor-crm/markets/views.py:
```diff
@@ -1,47 +1,239 @@
+    print("sector", sector)
+    print("industry", industry)
+
+    template = """Context: {question}
+                """
+    input_question = f"""You are a web application from which insights on Stocks is obtained. \n
+                The user interface of this web application contains a list of stocks on the Ghana Stock Exchange. \n
+                Each time a user of the web application selects a ticker, the equivalent company name,\
+                sector, industry, and market capitalization is passed to you.  \n
+                Your task is to generate a description that is displayed to the user. \n
+                Note, the description provided is displayed on web application that is accessed \
+                by professionals within the Ghanaian investment industry so the language should be professional \
+                and not conversational. Also, no need to apologise, just get straight to the point. \n
+                Your response should always begin with the company name provided to you. \n
+                Company name: {stock_name} \n
+                Company ticker: {ticker_info} \n
+                Company sector: {sector} \n
+                Company industry: {industry} \n
+                Company market capitalization {market_cap}"""
```
I put this in a file and read it so this is cleaner.

dswired commented on fin-advisor-crm/markets/views.py:
```diff
@@ -1,47 +1,239 @@
+    except DailyPublicEquityTrades.DoesNotExist:
+        return JsonResponse({"error": "Ticker not found"}, status=404)
+
+    trend_input_question = f"""You are a web application from which insights on Stocks is obtained. \n
+                The user interface of this web application contains a list of stocks on the Ghana Stock Exchange. \n
+                Each time a user of the web application selects a ticker, a corresponding list of stock prices \
+                will be passed. Your task is to generate interesting insights of the trend based on the list \
+                of prices. Bear in mind, the stock price is open source. \n
+                Company name: {stock_name} \n
+                Company sector: {sector} \n
+                Company industry: {industry} \n
+                Company historical prices: {price_labels} \n
+                """
```
Same as above...best read this from a file.

dswired commented on fin-advisor-crm/static/js/graphs.js:
```diff
@@ -1,14 +1,37 @@
-function loadChart(elementId, endPoint) {
+let chart = null;
+function loadChart(
+    selectedTicker, 
+    chartendPoint, 
+    chartElementId,
+    selectedDate,
+    ticker_tag,
+    date_tag,
+    ) 
+    {
+    const endPoint = chartendPoint;
     console.log("Fetching data from "+ endPoint);
-    fetch(endPoint)
+    console.log("Selected ticker "+ selectedTicker + "for date" + selectedDate);
+
+    const payload = {
+        // 'equityticker' : selectedTicker,
+        [ticker_tag] : selectedTicker,
+        [date_tag] : selectedDate
+    }
+
+    fetch(endPoint, {
+        method: "POST",
+        body: JSON.stringify(payload),
+        headers: { 'Content-Type': 'application/json' }        
+    })
```
`graphs.js` is meant to be used by all pages in the application so since this is very specific to your page put it in a different .js file and call that in your html file.

kayfoh replied:
This function is a generic function that can be implemented by any app within the project.

dswired commented on fin-advisor-crm/static/js/main.js:
```diff
@@ -1,9 +1,46 @@
+        console.error("Error fetching summary card data:", error);
+        // Handle error gracefully (e.g., display an error message to the user)
+    });
+}
```
`main.js` is meant to be used by all pages in the application so since this is very specific to your page put it in a different .js file and call that in your html file.

kayfoh committed:
updated code based on comments from pull request

kayfoh committed:
Page multi-function

-Implemented toggle card to switch between price and volume
-Condition trend data pull on toggle selection on UI
-Moved updateSummaryCard function to separate js file market_utils
-Switched from daily trend view to quarterly trend view

dswired commented:
@kayfoh Please resolve conflicts in this pr to have this merged.
